src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,365 @@
+/**
+ * @file linear_svm_function_impl.hpp
+ * @author Shikhar Bhardwaj
+ * @author Ayush Chamoli
+ *
+ * Implementation of the hinge loss function for training a linear SVM with the
+ * parallel SGD algorithm
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_FUNCTION_IMPL_HPP
+#define MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_FUNCTION_IMPL_HPP
+
+#include <mlpack/core/math/make_alias.hpp>
+#include <mlpack/core/math/shuffle_data.hpp>
+
+// In case it hasn't been included yet.
+#include "linear_svm_function.hpp"
+
+namespace mlpack {
+namespace svm {
+
+template <typename MatType>
+LinearSVMFunction<MatType>::LinearSVMFunction(
+  const MatType& dataset,
+  const arma::Row<size_t>& labels,
+  const size_t numClasses,
+  const double lambda) :
+  dataset(math::MakeAlias(const_cast<MatType&>(dataset), false)),
+  numClasses(numClasses),
+  lambda(lambda)
+{
+  // Initialize the parameters to suitable values.
+  initialPoint = InitializeWeights();
+
+  // Calculate the label matrix.
+  GetGroundTruthMatrix(labels, groundTruth);
+}
+
+/**
+ * Initializes parameter weights to random values taken from a scaled standard
+ * normal distribution. The weights cannot be initialized to zero, as that will
+ * lead to each class output being the same.
+ */
+template <typename MatType>
+const arma::mat LinearSVMFunction<MatType>::InitializeWeights()
+{
+  return InitializeWeights(dataset.n_rows, numClasses);
+}
+
+template <typename MatType>
+const arma::mat LinearSVMFunction<MatType>::InitializeWeights(
+        const size_t featureSize,
+        const size_t numClasses)
+{
+  arma::mat parameters;
+  InitializeWeights(parameters, featureSize, numClasses);
+  return parameters;
+}
+
+template <typename MatType>
+void LinearSVMFunction<MatType>::InitializeWeights(
+        arma::mat &weights,
+        const size_t featureSize,
+        const size_t numClasses)
+{
+  // Initialize values to 0.005 * r. 'r' is a matrix of random values taken from
+  // a Gaussian distribution with mean zero and variance one.
+  weights.randn(numClasses, featureSize);
+  weights *= 0.005;
```

Comment:
Fiven the simplicity of the initialization here, I would suggest entirely removing the `InitializeWeights()` method and its two overloads, and inline it into the constructor where it's only used once.